### ISSUE TITLE
Hierarchical discrete action space

### ISSUE THREAD

**Issue by cjfcsjt** [iffiX/machin#23]
Hi! Thanks for your excellent work!
I tried several RL frameworks based on PyTorch, machin is one of the few libraries that discusses hybrid action space.

I'm diving into a complex environment which is a hierarchical action space problem. I hope you could give me some advice!

To explain the meaning of hierarchical action space more clearly, here is an example in the paper [Generalising Discrete Action Spaces with Conditional Action Trees](https://arxiv.org/pdf/2104.07294.pdf). Figure2 in the paper shows that the actions are decomposed as an action tree. One should first select the first level actions, then select the second level actions. The action space of the first level is 3 and the action space of the second level depends on the first level.

I try to give one possible solution to solve this: 

1. change the transition part:
```python
transition = { "state": {"some_state": old_state} , ...  } # old
transition = {"state": {"some_state": old_state, "valid_actions": valid_action_set } , ... }
```
here, the "valid_actions" contains all of the possible second-level actions based on the first-level action.

2. change the agent sampling(explore) flow:
```python
state = env.reset() # state have key words "some_state" and "valid_actions", here, for initial, the valid_actions are pre-defined manually. For example, we choose the first-level action 'use', and then the valid_actions will be 'food'
while not done:
    action2 = agent.act2(state) # choose action2 from valid action set. Here, action2 is 'food'
    action1 = agent.act1(state) # choose action1 which denotes the first level action space. Here, suppose action1 is 'move'.
    env.first_level(action1) # tell env, the valid action set in next step is 'up, down, right, left' under the 'move' branch
    next_state, reward, done = env.step(action2) # next_state have key words "some_state" and "valid_actions", the valid_actions have 'up, down, right, left'.
```

3. change the actor network agent.act2 to:
```python
class act2(nn.Module):
  def forward(self, state = {'some_state': <...> , 'valid_actions': <...>}  ):
    state, second_level_actions = state.state, state.valid_actions
    ....  # calculate the similarity between state and valid actions, and output logits
    return (...), None
```
Do you think this solution is reasonable? Is there any better way to support such a conditional hierarchical action space?

**Comment by iffiX** [comment]
Hi, the answer below will work for the current highest version [0.4.2](https://github.com/iffiX/machin/tree/v0.4.2) and newer versions to be released.

---
### A small correction
To avoid causing confusion for future readers, I modified some details in step 2 of your comment:
> 2. change the agent sampling(explore) flow:
>
>```python
>state = env.reset() # state have key words "some_state" and "valid_actions", here, for initial, the valid_actions are pre->defined manually. For example, we choose the first-level action 'use', and then the valid_actions will be 'food'
>while not done:
>    action2 = agent.act2(state) # choose action2 from valid action set. Here, action2 is 'food'
>    action1 = agent.act1(state) # choose action1 which denotes the first level action space. Here, suppose action1 is 'move'.
>    env.first_level(action1) # tell env, the valid action set in next step is 'up, down, right, left' under the 'move' branch
>    next_state, reward, done = env.step(action2) # next_state have key words "some_state" and "valid_actions", the valid_actions have 'up, down, right, left'.
>```
to the newer version below, in the context of figure 2 of the paper [Generalising Discrete Action Spaces with Conditional Action Trees](https://arxiv.org/pdf/2104.07294.pdf). 

To my understanding, I think the paper is flattening the conditional action tree (CAT) and select an action level by level, therefore, instead of generating leaves by combinatorial rules (4 + 4 + 1 = 9 leaves here), for each level, they are using the maximum set size in each level instead (max(3) = 3, max(4, 4, 1) = 4), according to the Table 1 of the paper. (So logits in each level could possibly mean different things here, eg, if you choose action "use" in C0, then the mask of the second level (4 logits) should be [1, 0, 0, 0] if "use food" is permitted in the environment, or [0, 0, 0, 0] if it is not permitted).

I am also not sure about what they mean by **full conditional masking**, or referred to as **CAT_CD** in contrast to the **collapsed conditional masking**, referred to as **CAT_CL** in section V-D of the paper, but that's not the main concern as long as your sampled action logits (p(c_k | m_k, s)) and sampled mask logits (p(m_k+1 | c_k)) has the same size in each level of CAT, **independent of your action** (So even if you choose action "use" in C0, the action logits and mask logits will be of size 4 in C1, with the last 3 logits always be 0 since they don't mean anything valid), I will explain this necessity in the next section. 
![Figure 2](https://user-images.githubusercontent.com/17508457/129011319-ae1dd723-8f6d-493e-8a2b-25dab0117e42.png)

2. change the agent sampling(explore) flow:
```python
state = env.reset() # state have key words "some_state" and "valid_actions", here, for initial, the valid_actions are pre-defined manually. For example, we choose the first-level action 'use', and then the valid_actions will be 'food'
while not done:
    # Agent is some type of algorithm framework, like PPO, DDPG, etc. Depending on the framework, result of "agent.act" could be different.
    # First choose action1 which denotes the first level action space. Here, suppose action1 mean 'move', represented by an array [1, 0, 0] in C0.
    # Then choose action2 from valid action set. Here, action2 could mean 'down', represented by array [0, 1, 0, 0] in C1.
    action1,  action2 = agent.act(state) 
    env.first_level(action1) # tell env, the valid action set in next step is 'up, down, right, left' under the 'move' branch
    next_state, reward, done = env.step(action2) # next_state have key words "some_state" and "valid_actions", the valid_actions have 'up, down, right, left'.
```

### Your solution
**Absolutely reasonable**, the framework is designed to work like this, just a few more corrections:
 ```python
class SomeNetwork(nn.Module):
  def forward(self, some_state , valid_actions):
    # deal with each level of actions here, like action1, action2, etc.
    return (...)
```
First, the "state" and "action" attributes are dicts and should not contain the same keys, as their keys will be concatenated and passed to your model directly. Details are explained [here](https://machin.readthedocs.io/en/latest/tutorials/data_flow_in_machin.html), though some parts of this document need to be updated.

Second, the "valid_actions" should be a tensor, or an object convertible to a tensor, they should also be concatenatable in dimension 0, which means they should have the same size in other dimensions, otherwise an exception will be thrown by:
https://github.com/iffiX/machin/blob/faf7c72f9cb88d028519e5af38ea8f8f3e886c88/machin/frame/buffers/buffer.py#L326

That's why you need to make sure that your "valid_actions" for CAT should be the same size for any step in any episode, if they are not concatenatable, you can only set "concatenate" = False in update calls to the framework (agent here), and deal with lists of unconcatenated samples for each attribute in your network.

**Comment by cjfcsjt** [comment]
Thank you very much for your reply!

#### about the paper
The paper tries to utilize the mask mechanism to cope with the "varying action space" problem. I found two possible ways in the literature to cope with such a problem. 

The first one is the masking mechanism, Specifically, the actor gives a probability vector with a fixed maximum size, we give extremely small values to all the 'illegal' actions. This would involve a step in the Network that sets the probability outputs of illegal actions to 0, and re-normalizes the rest to sum up to 1 again. The second one is to give the actor-network the actual valid action set and let the network to calculate the similarity between the state and each action in action set. 

I believe that both two methods need to add an item in "state" dict (e.g. trainsition['state'] = {'state':state, 'mask': mask}). For the first one, we add mask in "state" dict while for the second one, we add "valid action set" in "state" dict. Which one do you think is better? Or these two methods are both reasonable?

#### step 2 modification 
I have read your modification in step2. You utilize one actor-network with two heads instead of two actor-networks. Since I am a novice in RL, this modification sounds novel to me. I found another issue #20 which calls such a structure the multi-discrete action space. I will try the solution in #20 later.

> First, the "state" and "action" attributes are dicts and should not contain the same keys, as their keys will be concatenated and passed to your model directly. Details are explained here, though some parts of this document need to be updated.

I read the 'Data flow in Machin' part in doc. I guess that you are talking about the 'safe call'. For example, we can custom a transition like below:
```python
transition = {
  "state": {"cat": cat, "rabbit": rabbit, "fish": fish},
  "action": {"action1": action1, "action2":action2, "action3":action3},
  "next_state": {"cat": cat_new, "rabbit": rabbit_new, "fish": fish_new},
  "reward": 1.0
   "terminal": False
}
```
We also have two actor-networks like below:
```python
class Actor-Network1(nn.Module):
 def forward(self, cat, fish):
   # ...
   return action 1

class Actor-Network2(nn.Module):
 def forward(self, cat, fish):
   # ...
   return action2, action3
```
Then, in my understanding, the "state" dict in transition is indeed the **input** of the network while the "action" can be defined as the **output** of the network. "Safe Call" aims to assign the right arguments towards the forward function. If you pass the {"cat": cat, "rabbit": rabbit, "fish": fish} to Actor-Network1 by safe-call, then it will automatically pass the cat and fish into Actor-Network1 and ignore the rabbit.

When I write down the above paragraph, I realized that I can combine the Actor-Network1 and Actor-Network2 into on Actor-Network as you mentioned before. I think such a combination leads to only a small modification in update() function as mentioned in #20. However, If we use multiple Actor-Network, then, in update(), we need to calculate the loss and call loss.backward() for each actor-network separately. 

As a result, the combined Actor-Network will be 
```python
class Actor-Network1(nn.Module):
 def forward(self, cat, fish):
   #  a sub-network take cat and finish as input.
   #  three output heads output action1, action2 and action3 respectively 
   return action1, action2, action3
```
Is my understanding correct？

By the way, there is another paper talk about the hierarchical action space which may give you some new insights: [Hybrid Actor-Critic Reinforcement Learning in Parameterized Action Space](https://www.ijcai.org/proceedings/2019/0316.pdf)

**Comment by iffiX** [comment]
Hi,
As long as your actor network gives a probability vector with a fixed size, both solutions will work, and they are both reasonable depending on your model and method.

Your understanding of the safe call mechanism is correct, and it is also recommended to encapsulate sub modules into one network, which serves the "Actor" functionality, so readers of your code will be more clear about the role of each network module. If you need to customize the update() call, just inherit from the framework and overload it.

The new paper provides some good insight on the hybrid action space, I already understand the method used here, it's just that the previous paper you mentioned did not explain some details accurately, thanks for the new paper, it will certainly benefit others reading this issue!
